Provider/src/main/java/com/spotify/confidence/Confidence.kt

```diff
@@ -40,34 +50,50 @@ class Confidence internal constructor(
     }
 
     internal suspend fun resolve(flags: List<String>): Result<FlagResolution> {
-        return flagResolver.resolve(flags, getContext().openFeatureFlatten())
+        return flagResolver.resolve(flags, getContext())
     }
 
     internal fun apply(flagName: String, resolveToken: String) {
         flagApplier.apply(flagName, resolveToken)
     }
 
+    @Synchronized
     override fun putContext(key: String, value: ConfidenceValue) {
-        contextMap[key] = value
+        val map = contextMap.value.toMutableMap()
+        map[key] = value
+        contextMap.value = map
     }
 
+    @Synchronized
     override fun putContext(context: Map<String, ConfidenceValue>) {
-        contextMap += context
+        val map = contextMap.value.toMutableMap()
+        map += context
+        contextMap.value = map
     }
 
-    override fun setContext(context: Map<String, ConfidenceValue>) {
-        contextMap = context.toMutableMap()
+    @Synchronized
+    internal fun putContext(context: Map<String, ConfidenceValue>, removedKeys: List<String>) {
+        val map = contextMap.value.toMutableMap()
+        map += context
+        for (key in removedKeys) {
+            map.remove(key)
+        }
+        this.removedKeys.addAll(removedKeys)
+        contextMap.value = map
     }
 
+    @Synchronized
     override fun removeContext(key: String) {
+        val map = contextMap.value.toMutableMap()
+        map.remove(key)
         removedKeys.add(key)
```

Comment:
In Swift, the removed keys are rather stored in Confidence: https://github.com/spotify/confidence-openfeature-provider-swift/blob/0d1cefdeb766a3d24c7d05be5f834d8855f271f3/Sources/Confidence/Confidence.swift#L101C24-L101C42
Let's align on one approach

Reply:
it's the same, this class is confidence. the link in swift the `confidence` variable is something that comes from the confidence. maybe the naming is a bit confusing over there.